feat: introduce `libp2p-stream`

thomaseizinger opened the pull request:
## Description

For a while now, `rust-libp2p` provided the `request-response` abstraction which makes it easy for users to build request-response based protocols without having to implement a `NetworkBehaviour` themselves. This PR introduces an alpha version of `libp2p-stream`: a `NetworkBehaviour` that directly gives access to negotiated streams.

In addition to complementing `request-response`, `libp2p-stream` also diverges in its design from the remaining modules by offering a clonable `Control` that provides `async` functions.

Resolves: #4457.

<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

## Notes & open questions

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Add basic boilerplate for stream example

thomaseizinger committed:
Write the demo application

thomaseizinger committed:
Introduce `PeerControl`

thomaseizinger committed:
Initial implementation

mxinden commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,140 @@
+    // Deal with incoming streams.
+    // Spawning a dedicated task is just one way of doing this.
+    // libp2p doesn't care how you handle incoming streams but you _must_ handle them somehow.
+    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep in processing them.
```
I see two potential mechanisms for handling incoming streams:

1. User provides a handler
2. `stream::Behaviour` provides a `Stream` of streams

Intuitively I opted for (2). The simplicity below also supports (2) to be the better option.

Agreed?

thomaseizinger replied:
Yes. I think (2) is the better API because it gives the user more flexibility.

mxinden commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,140 @@
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::EnvFilter;
+
+const PROTOCOL: StreamProtocol = StreamProtocol::new("/my-ping-protocol");
```
To prevent confusion with the existing `libp2p-ping`, how about implementing an _echo_ protocol in this example? On the wire, this is the same thing. But using a different name prevents users from thinking that this is **the* libp2p way to ping.

thomaseizinger replied:
Should the protocol then be to echo a length-prefixed message?

mxinden replied:
Fine by me. Or the server reads until stream closing and the echoes the message.

mxinden commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,140 @@
+    let mut peer_control = control.peer(peer).await.unwrap();
+
+    loop {
+        tokio::time::sleep(Duration::from_secs(1)).await; // Wait a second between pings.
```
Could this sleep as well be moved to the bottom of the `loop`? Or is it necessary to wait for the dial. Moving it to the bottom in my eyes makes this example more focused.

thomaseizinger replied:
Then all `continue` statements would immediately try again. I wanted the delay be effective for those as well.

mxinden commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,140 @@
+
+/// A very simple, `async fn`-based connection handler for our custom ping protocol.
+async fn connection_handler(peer: PeerId, mut control: stream::Control) {
+    let mut peer_control = control.peer(peer).await.unwrap();
```
Why do we need the additional indirection of a peer control? Why not allow a user to create a stream off of the `stream::Control` directly?

mxinden replied:
Without the indirection, garbage collection is also easier, right? E.g. what if all `Control`s of a protocol are dropped but a `PeerControl` for the user protocol still exists?

thomaseizinger replied:
That is what I had at first but I think it makes upholding backpressure complicated: The message contains the `PeerId` so I need to read from the channel before I know which channel to a connection handler would want the message but that channel might be full! So in order to uphold backpressure, I now need to buffer this message and stop polling the channel until I can send the other one. As a result, the slowest connection will hold back all others.

For this protocol, all the handler needs to do is open a stream so one might argue, it doesn't matter that much but I just don't find it a very clean solution.

Do you have any ideas on how else to solve this?

An unrelated aspect that also made me go for the `PeerControl` was that it could be nice extension point if one wants to add the ability to e.g. control the dialing by adding addresses to it. However, I opted not to implement that for simplicity and I am not convinced it is something that this behaviour should deal with.

thomaseizinger replied:
Something else I just came across is connection keep-alive. My current plan is to keep connections alive while `PeerControl`s exist. Do you think that is a good idea? On the one hand, it gives users _some_ control over managing connections. On the other hand, there is built-in functionality to just dial the remote peer when we want to send them a message.

thomaseizinger replied:
Doing the keep-alive is actually somewhat tricky and with the receiver's being shared across all connections, it isn't very clean because now _all_ connections would be kept alive.

I think I'll remove that again.

The point above still stands though. I don't know how else to model it cleanly other than with the indirection of `PeerControl`. Ideas welcome!

mxinden commented on examples/stream/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+tracing = "0.1.37"
+tracing-subscriber = { version = "0.3", features = ["env-filter"] }
+rand = "0.8"
+flume = "0.11.0"
```
If I understand correctly, we only need `flume` to do work stealing. Is that correct?

I would argue that the benefit of work-stealing does not warrant the additional dependency. Instead I would suggest a `futures::channel::mpsc` to each `ConnectionHandler`. The channels each can have a capacity of `0` (effectively `1` given that each sender gets a slot). the `NetworkBehaviour` iterates through all channels and sees if one has capacity.

thomaseizinger replied:
Going to resolve this in favor of the discussion above. It is not just about work-stealing although I do find that it plays well together. I am open to removing `flume` again if we can solve the above in a clean way. `flume` is also a very small dependency so I don't think it matters much.

mxinden commented on examples/stream/src/lib.rs:
```diff
@@ -0,0 +1,517 @@
+#[derive(Debug)]
+pub struct AlreadyRegistered;
+
+// TODO: On `Drop`, we need to de-register the protocol.
```
On drop of all instances of `Control`. Correct?

Though given that this carries a `mpsc::Sender`, one might as well do the garbage collection when the `Receiver` end closes.

thomaseizinger replied:
I just thought about this some more and we don't actually need to do this. The `Control` is only for _opening_ new streams. It is a perfectly legit use of this module to immediately drop the `Sender` if your application is never going to open a stream for a certain protocol. Plus, I actually think you should be able to obtain a new `Control` at any point.

What should de-register the protocol is dropping `IncomingStreams`. Thinking more about this, it seems like a better API to decouple these two things. One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.

mxinden replied:
> One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.

Reading your first paragraph, I had the same thought.

Maybe separate methods for all 3 options (in, out, in and out)? Or an `enum` as a parameter to the method?

thomaseizinger replied:
> > One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.
> 
> Reading your first paragraph, I had the same thought.
> 
> Maybe separate methods for all 3 options (in, out, in and out)? Or an `enum` as a parameter to the method?

Have a look at the latest version. I quite like the simplicity of a dedicated `new_control` and a `accept` function. The names could be improved. Do you have any ideas?

mxinden commented on examples/stream/src/lib.rs:
```diff
@@ -0,0 +1,517 @@
+
+pub struct Upgrade {
+    supported_protocols: Vec<StreamProtocol>,
+}
```
Maybe worth abstracting into a `libp2p_swarm::upgrade::ReadyMany`?

thomaseizinger replied:
I am a bit torn. I think it could be useful but at the same time, I don't think it is a step in the right direction to expose utilities from `libp2p-swarm`.

mxinden reviewed (commented):
This is great! The implementation is simple, the example is even simpler.

I have some suggestions but overall, from an architecture perspective, this looks good to me.

thomaseizinger committed:
Split code up into modules

thomaseizinger committed:
Split `register` into `new_control` and `accept`

thomaseizinger committed:
De-register protocol on drop of `IncomingStreams`

thomaseizinger committed:
Add TODO

thomaseizinger committed:
Import `Pin`

thomaseizinger committed:
Introduce `libp2p-stream`

thomaseizinger committed:
Improve docs

thomaseizinger committed:
Rename error

thomaseizinger committed:
Remove obsolete TODO

thomaseizinger committed:
Make tests pass

thomaseizinger committed:
Add failing test

thomaseizinger committed:
Implement some keep-alive logic

thomaseizinger committed:
Use flume rendezvous channel for `IncomingStreams`

The idea is that a rendezvous channel will wake up the connection handler
and allows us to directly close the connection if nothing else is
using it.

thomaseizinger committed:
`IncomingStreams` should not keep connection alive

thomaseizinger committed:
Add TODOs

thomaseizinger committed:
Remove `PeerControl` API

thomaseizinger committed:
Update readme

thomaseizinger committed:
Change example to echo protocol

mxinden commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,153 @@
+    // Deal with incoming streams.
+    // Spawning a dedicated task is just one way of doing this.
+    // libp2p doesn't care how you handle incoming streams but you _must_ handle them somehow.
+    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep in processing them.
```
```suggestion
    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep up processing them.
```

mxinden commented on protocols/stream/src/behaviour.rs:
```diff
@@ -0,0 +1,287 @@
+            Some(sender) => {
+                tracing::debug!("Returning sender to existing connection");
+
+                sender.clone()
```
This will expand the size of the channel and thus break backpressure, no?

thomaseizinger replied:
Yes but a user would have to to out of their way to achieve that by also cloning the `Control`.

If a user calls `open_stream` in a loop, we will still uphold backpressure and only open one stream at a time.
If they clone the `Control` or don't immediately await the future then a new `Sender` will be created for each one. I think that is fairly intuitive actually. I think the only concern here is increased memory-consumption and doing any of the above two suggests to the user that more memory will be consumed.

Overall, I think this is acceptable. Do you see any other problems?

thomaseizinger replied:
Even awaiting multiple futures doesn't work because we take `&mut self` so the user MUST clone the `Sender`. In other words, we are merely passing on the behaviour of an `mpsc::Sender` to `Control`.

We should probably document that though.

mxinden reviewed (commented):
New design using a `Mutex` looks good to me. One note though:

thomaseizinger committed:
Update examples/stream/src/main.rs

Co-authored-by: Max Inden <mail@max-inden.de>

thomaseizinger commented on protocols/stream/src/behaviour.rs:
```diff
@@ -0,0 +1,287 @@
+    ///
+    /// A [`Control`] only deals with the _outbound_ side of a protocol.
+    /// To accept inbound streams for a protocol, use [`Behaviour::accept`].
+    pub fn new_control(&self, protocol: StreamProtocol) -> Control {
```
I think I want to move the protocol argument to `Control::open_stream`. It seems unnecessarily limiting to force the user to create multiple `Control`s.

thomaseizinger committed:
Make `Control` independent of protocol

thomaseizinger committed:
Move `accept` to `Control`

thomaseizinger committed:
Add test for dial error

thomaseizinger committed:
Improve docs

thomaseizinger committed:
Don't unwarp poisoned lock

thomaseizinger committed:
Introduce `control.rs` module

thomaseizinger committed:
Introduce `shared` module

thomaseizinger committed:
Add backpressure docs

thomaseizinger committed:
Add `Default` impl for `Behaviour`

thomaseizinger committed:
Add metadata to manifest

thomaseizinger committed:
Fix doc links

thomaseizinger committed:
Merge branch 'master' into feat/stream-behaviour

thomaseizinger committed:
Add version entry

thomaseizinger committed:
Undo changes in swarm

thomaseizinger committed:
Add changelog file

mxinden commented on protocols/stream/README.md:
```diff
@@ -0,0 +1,69 @@
+You must continuously poll it to make progress.
+In the example above, this taken care of by using the [`StreamExt::next`](futures::StreamExt::next) helper.
+
+Internally, we will drop streams if your application falls behind in processing these incoming streams, i.e. if whatever loop calls `.next()` is not fast enough.
```
Nit pick, given that new streams are still instantiated, but then dropped, this is not really backpressure, but instead tail drop. Maybe change the heading to `### Resource management`?

thomaseizinger replied:
Thanks! I actually didn't mean to imply that this sentence is backpressure! I'll revise the wording to make this clear :)

mxinden reviewed (approved):
Looking forward for this to land! I expect this to be useful for many, especially those implementing a protocol on top of libp2p for the first time.

thomaseizinger committed:
Correct title to resource management

rkuhn commented on protocols/stream/src/handler.rs:
```diff
@@ -0,0 +1,165 @@
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(
```
When is this hook called? I’m asking because the supported protocols may change during the lifecycle of a connection.

thomaseizinger replied:
This is called every time a new inbound stream arrives. There is a test that checks that upon dropping `IncomingStreams`, the protocol is deregistered.

rkuhn commented on examples/stream/src/main.rs:
```diff
@@ -0,0 +1,154 @@
+            event => tracing::trace!(?event),
+        }
+    }
+}
```
looks nice and clear to me

mergify[bot] committed:
Merge branch 'master' into feat/stream-behaviour
